core/shared/src/main/scala/scalaz/zio/Managed.scala

```diff
@@ -0,0 +1,82 @@
+package scalaz.zio
+
+sealed abstract class Managed[+E, +R] { self =>
+  def use[E1 >: E, A](f: R => IO[E1, A]): IO[E1, A]
+
+  def use_[E1 >: E, A](f: IO[E1, A]): IO[E1, A] =
+    use(_ => f)
+
+  def map[R1](f0: R => R1): Managed[E, R1] =
+    new Managed[E, R1] {
+      def use[E1 >: E, A](f: R1 => IO[E1, A]): IO[E1, A] =
+        self.use(r => f(f0(r)))
+    }
+
+  def flatMap[E1 >: E, R1](f0: R => Managed[E1, R1]) =
+    new Managed[E1, R1] {
+      def use[E2 >: E1, A](f: R1 => IO[E2, A]): IO[E2, A] =
+        self.use { r =>
+          f0(r).use { r1 =>
+            f(r1)
+          }
+        }
+    }
+
+  def *>[E1 >: E, R1](ff: Managed[E1, R1]): Managed[E1, R1] =
+    flatMap(_ => ff)
+
+  def <*[E1 >: E, R1](ff: Managed[E1, R1]): Managed[E1, R] =
+    flatMap(r => ff.map(_ => r))
+
+  def seqWith[E1 >: E, R1, R2](ff: Managed[E1, R1])(f: (R, R1) => R2): Managed[E1, R2] =
```

Comment:
Can we make the `par` variants safe as well?

Reply:
To do this, I think we'll need to change the encoding to keep the resource's IO and its cleanup action in the Managed instance. Otherwise we won't have access to `par` on the acquire IO.

I'll fiddle with that and see how that looks.

Reply:
Ok, I'm pretty sure that to do this we'll need to encode Managed as an algebra like in cats-effect.

I tried several encodings:
- Abstract methods `def acquire: IO[E, R]` and `def release(r: R): IO[Nothing, Unit]`. The problem with this encoding is that `R` is now in contravariant position, so we can't implement `map`.
- Abstract method `def resource: IO[E, (R, IO[Nothing, Unit])]`. This time we won't be able to implement flatMap properly and keep the resource scoping right.
- Abstract methods `def use` and `def resource` - this works but has weird artefacts; the Managed instance produced in `map` will ignore its `resource` member and use the resource from the previous Managed

On the other hand, if users would like to acquire resources in parallel, they can combine the IO values outside of Managed; e.g.:
```Managed(acquire1 parSeq acquire2)(rs => releaseAll(rs._1, rs._2))```

So this might not be worth it.

Reply:
OTOH, an algebra for Managed might enable other interpreters that will acquire resources with different scoping rules. E.g., in streams.

Happy to implement if you think this is the right way.